dallinger/mturk.py

```diff
@@ -32,124 +31,186 @@ class QualificationNotFoundException(MTurkServiceException):
 
     @cached_property
     def mturk(self):
-        """Cached MTurkConnection"""
-        if not self.aws_access_key_id or not self.aws_secret_access_key:
-            raise MTurkServiceException('AWS access key and secret not set.')
-        login_params = {
-            'aws_access_key_id': self.aws_access_key_id,
-            'aws_secret_access_key': self.aws_secret_access_key,
-            'host': self.host
-        }
-        return MTurkConnection(**login_params)
+        session = boto3.session.Session(
+            aws_access_key_id=self.aws_key,
+            aws_secret_access_key=self.aws_secret,
+            region_name=self.region_name
+        )
+        return session.client(
+            'mturk',
+            endpoint_url=self.host,
+            region_name=self.region_name,
+        )
 
     @property
     def host(self):
         if self.is_sandbox:
-            return self.sandbox_mturk_server
-        return self.production_mturk_server
+            template = u'https://mturk-requester-sandbox.{}.amazonaws.com'
+        else:
+            template = u'https://mturk-requester.{}.amazonaws.com'
+        return template.format(self.region_name)
+
+    @property
+    def legacy_host(self):
+        """Setting endpoints for REST notifications is not supported by the
+        latest version of the AWS MTurk API, and this is the endpoint boto3
+        communicates with. As a workaround, we use the old endpoint for this
+        one feature.
+        """
+        if self.is_sandbox:
+            return 'mechanicalturk.sandbox.amazonaws.com'
+        return 'mechanicalturk.amazonaws.com'
 
     def check_credentials(self):
         """Verifies key/secret/host combination by making a balance inquiry"""
-        return bool(self.mturk.get_account_balance())
+        try:
+            return bool(self.mturk.get_account_balance())
+        except NoCredentialsError:
+            raise MTurkServiceException("No AWS credentials set!")
+        except ClientError:
+            raise MTurkServiceException("Invalid AWS credentials!")
+        except Exception as ex:
+            raise MTurkServiceException(
+                "Error checking credentials: {}".format(ex.message)
+            )
 
     def set_rest_notification(self, url, hit_type_id):
-        """Set a REST endpoint to recieve notifications about the HIT"""
-        all_events = (
-            "AssignmentAccepted",
-            "AssignmentAbandoned",
-            "AssignmentReturned",
-            "AssignmentSubmitted",
-            "HITReviewable",
-            "HITExpired",
-        )
-
-        return self._is_ok(
-            self.mturk.set_rest_notification(hit_type_id, url, event_types=all_events)
+        """Set a REST endpoint to recieve notifications about the HIT
+        The newer AWS MTurk API does not support this feature, which means we
+        cannot use boto3 here. Instead, we make the call manually after
+        assembling a properly signed request.
```

Comment:
Looks like the officially supported alternative would be to send notifications via SNS to an HTTP endpoint? This seems okay for now but I guess we will need to move to that at some point in the future.

Reply:
That's my understanding also: email or SNS are now the supported notification options.